### ISSUE TITLE
Guard/prevent dangerous actions like API token deletion better

### ISSUE THREAD

**Issue by rugk** [openstreetmap/openstreetmap-website#3551]
## Problem

The API token on the website is only guarded by one `confirm()` (alert/prompt) prompt.
See this screenshot:

![grafik](https://user-images.githubusercontent.com/11966684/168497274-2483835f-c5b6-43a2-9fba-0f06174eba19.png)

If you accidentally click "Delete", this has potentially _severe consequences_. All your users of the app are force logged-out and need to login. _And_ you likely need to update the code and ship an update of your application, which again may take some time. This all results in a very bad UX impediment for your users. E.g. users could be blocked for days from using your app. (until they get that update.)
For (potential) apps that make money based on the features they offer, this can result in a sales loss, in the worst case.

Note this happened IRL: https://github.com/streetcomplete/StreetComplete/issues/4028#issuecomment-1126990804

### Variant 1: Undo

UX wise I'd propose two common solutions (you may keep the current dialog, or replace them by that new ones):
* add an “Undo” feature and do not delete the token right away, but after a “cooldown period” in which users can think whether they really wanted to do what they wanted…

The thing one has to keep in mind is that, in our case, we also need to be _fast_ deleting the API token, because the reason for deletion may also be that your token was stolen or so and someone already misused it and is doing edits etc.

### Screenshots
![grafik](https://user-images.githubusercontent.com/11966684/168497556-ce2ed947-7066-479c-ab49-3fddaddf38f0.png)

```html
<div class="flash warning row mx-0 p-3 align-items-center">
    <div class="col-auto">
      <picture>
        <source srcset="/assets/notice-8562cd4b85576531716bcc1543e08f82c82b6aea8244393974b5f4f306f85897.svg" type="image/svg+xml">
        <img srcset="/assets/notice-8562cd4b85576531716bcc1543e08f82c82b6aea8244393974b5f4f306f85897.svg" src="/assets/notice-d058d399b5d6f7b617d3cf8ff36c71b1f83c51cd4de6dd8ef1a194a451d48236.png">
      </picture>
    </div>
    <div class="" style="margin-right: 10px;">Application API key will be deleted in 10s.</div>
  <a class="btn btn-outline-primary" href="/oauth2/applications/new">Undo</a>
</div>
````

Maybe better mentioning the application name:
![grafik](https://user-images.githubusercontent.com/11966684/168497640-d5683797-578f-42f2-8b93-fe4ea79eed0e.png)

Also, the seconds there should count down so one sees how much time is left.

### Variant 2: Paper bin or revoke and re-enable
* actually, that's why you may implement something like a ”paper bin” that allows you to restore stuff…
* or, in a better term, maybe just allow revoking/disabling access tokens and allow them to be re-enabled. Then, deleting should either be impossible, or they heavily restricted (again UX-wise, i.e. a small button, or to be sure, technically, by only allowing to delete the token 30 days after it has been disabled e.g.)
  * also, IMHO, you could also go the way of automatically deleting revoked tokens after 30 days or so if you want to clean-up things. (Just add a small text/warning then, that this will happen.)
  * if you want to keep the manual delete functionality, you should only offer it for revoked tokens
  
Revoking in such a way would have the advantage that it would be immediately (good for the theft scenario), while still allowing easy recover of the previous token.

### Screenshots

![grafik](https://user-images.githubusercontent.com/11966684/168497889-7073297d-31fd-4a93-938b-4cb5ff347bde.png)

Again here, to better highlight it what happened to that token, you may add a text there (you can also call it “disabled”), and obviously the delete button is then used for revoking instead (e.g., depending on whether/how you want to keep the deletion functionality):
![grafik](https://user-images.githubusercontent.com/11966684/168498115-48792cb0-a9d0-41ff-85a8-3a9b1d9a54f5.png)

```html
<tr>
  <td class="align-middle">
    <ul class="list-unstyled mb-0">
      <li><a href="/oauth2/applications/1115" style="text-decoration: css;text-decoration: line-through;">MyApp Two</a><span> (<b>Revoked</b>)</span></li>
        <li class="text-muted" style="text-decoration: line-through;">https://test-2.example</li>
    </ul>
  </td>
  <td class="align-middle">
    <ul class="list-unstyled mb-0">
    </ul>
  </td>
  <td class="align-middle">
    <a class="btn btn-outline-primary" href="/oauth2/applications/1115/edit" style="display: none;">Bearbeiten</a>
  </td>
  <td class="align-middle">
    <a data-confirm="Diese Anwendung löschen?" rel="nofollow" data-method="delete" href="/oauth2/applications/1115" class="btn btn-outline-info">Enable</a>
  </td>
</tr>
```

## Final notes

Of course, all approaches could be combined, including the current “confirm one”.

Needless to say BTW, in case you have other, similarly dangerous actions on your website, you should “UX-protect” them in the same way.

**Comment by tomhughes** [comment]
Sure, we could do all sorts of things, but realistically we don't have infinite development resources and we already ask for confirmation - spending a huge amount of time adding a feature that will probably never be used doesn't really seem very sensible when it already has quite reasonable protection against anybody that isn't absolutely determined to shoot themselves in the foot.

Please note that what you have tested here is not the actual case involved in https://github.com/streetcomplete/StreetComplete/issues/4028 as that was using OAuth 1 not OAuth 2 and that quite probably doesn't have any confirmation, but it's also a legacy protocol that apps should be moving away from.

**Comment by rugk** [comment]
Sure, I understand you have limited resources and it's obviously up to you what issues should be prioritized.
The thing is just that for dangerous user actions, the impact has to be evaluated here. And I guess we have a potentially damaging impact here – only marginally lower than “delete my user profile”. [The undo button is a well-known solution to that problem](https://ux.stackexchange.com/a/71961/123725).
Also, [there are other solutions you could take](https://uxmovement.com/buttons/how-to-design-destructive-actions-that-prevent-data-loss/), from using a good color coding to typing text (like GitHub's “delete account” feature – that could even be done with a simple `prompt` in JS). If we wanted more concrete/similar UX examples, one could look at how Google, Facebook or co handle tokens on their respective pages and how they can be deleted there.

So all in all, I see ways to improve here… some require more effort, some less. And IMHO, it's worth doing it here, since we had an incident and we know how we can systematically prevent such things in the future.

So as a concrete idea, what you could e.g. do is leaving this issue open and tagging it with “help wanted” so contributors could pick it up if they were to make UX improvements.

**Comment by westnordost** [comment]
There is an "Are you sure" dialog also for OAuth1 alright. 

I did not ignore the warning, I just thought I was deleting **the other** entry which I also named "StreetComplete" because I think this is the name that is displayed to users, so I was bound to use the same name.
I created that second entry because I needed to prompt the user for more permissions and did not know I could just edit the current app. I realized this some seconds after I created the second entry so I went on to again delete the just added entry, believing it was the last in the list. Turned out, it was the first in the list.

FWIW, from my point of view, from the point of view of the idiot that was not cautious enough, (any of) the following things would have prevented my mistake:
- sort the list of applications by id ascending
- (latest) in the "are you sure" dialog, show the number of users who authorized this app
